File: scala/src/main/scala/domain/Withdraw.scala
package bank.domain

import java.util.Date

class Withdraw(val amount: Double, operation: Date) extends Transaction(operation) {
  if(amount <= 0) throw new IllegalArgumentException("the number must be strictly positive.")

  def deposit: Double = 0.0

  def withdraw: Double = amount

  override def equals(that: Any): Boolean =
    that match {
      case that: Withdraw => this.amount.equals(that.amount)
      case _ => false
   }

  override def hashCode:Int = {
    val prime = 31
    var result = 1
    result = prime * result + amount.hashCode;
    return result
  }

  override def toString = s"Withdraw of $amount"

}

File: java/src/main/java/bank/Withdraw.java
package bank;

import java.time.LocalDate;
import java.util.Optional;

public class Withdraw extends Transaction {

	public Withdraw(float amount, LocalDate dateTime) {
		super(amount, dateTime);
	}

	@Override
	public Optional<Float> getDebit() {
		return Optional.of(amount);
	}

	@Override
	public Optional<Float> getCredit() {
		return Optional.empty();
	}

}
